ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/MergeCompactor.java

```diff
@@ -0,0 +1,224 @@
+      Map<Integer, List<Reader>> intermediateResult = getBucketFiles(conf, deltaDirPath, isMm);
+      intermediateResult.forEach((k, v) -> {
+        if (result.containsKey(k)) {
+          result.get(k).addAll(v);
+        } else {
+          result.put(k, v);
+        }
+      });
+    }
+    return result;
+  }
+
+  /**
+   * Collect the list of all bucket file paths, which belong to the same bucket Id. This method checks only one
+   * directory.
+   * @param conf hive configuration, must be not null
+   * @param dirPath the directory to be scanned.
+   * @param isMm collect bucket files fron insert only directories
+   * @throws IOException an error happened during the reading of the directory/bucket file
+   */
+  private Map<Integer, List<Reader>> getBucketFiles(HiveConf conf, Path dirPath, boolean isMm) throws IOException {
+    Map<Integer, List<Reader>> bucketIdToBucketFilePath = new HashMap<>();
+    FileSystem fs = dirPath.getFileSystem(conf);
+    FileStatus[] fileStatuses =
+            fs.listStatus(dirPath, isMm ? AcidUtils.originalBucketFilter : AcidUtils.bucketFileFilter);
+    for (FileStatus f : fileStatuses) {
+      final Path fPath = f.getPath();
+      Matcher matcher = isMm ? AcidUtils.LEGACY_BUCKET_DIGIT_PATTERN
+              .matcher(fPath.getName()) : AcidUtils.BUCKET_PATTERN.matcher(fPath.getName());
+      if (!matcher.find()) {
+        String errorMessage = String
+                .format("Found a bucket file which did not match the bucket pattern! %s Matcher=%s", fPath,
+                        matcher);
+        LOG.error(errorMessage);
+        throw new IllegalArgumentException(errorMessage);
+      }
+      int bucketNum = matcher.groupCount() > 0 ? Integer.parseInt(matcher.group(1)) : Integer.parseInt(matcher.group());
+      bucketIdToBucketFilePath.computeIfAbsent(bucketNum, ArrayList::new);
+      Reader reader = OrcFile.createReader(fs, fPath);
+      bucketIdToBucketFilePath.computeIfPresent(bucketNum, (k, v) -> v).add(reader);
+    }
+    return bucketIdToBucketFilePath;
+  }
+
+  /**
+   * Generate output path for compaction. This can be used to generate delta or base directories.
+   * @param conf hive configuration, must be non-null
+   * @param writeIds list of valid write IDs
+   * @param isBaseDir if base directory path should be generated
+   * @param sd the resolved storadge descriptor
+   * @return output path, always non-null
+   */
+  private Path getOutputDirPath(HiveConf conf, ValidWriteIdList writeIds, boolean isBaseDir,
+                                         StorageDescriptor sd) {
+    long minOpenWriteId = writeIds.getMinOpenWriteId() == null ? 1 : writeIds.getMinOpenWriteId();
+    long highWatermark = writeIds.getHighWatermark();
+    long compactorTxnId = Compactor.getCompactorTxnId(conf);
+    AcidOutputFormat.Options options = new AcidOutputFormat.Options(conf).writingBase(isBaseDir)
+            .writingDeleteDelta(false).isCompressed(false).minimumWriteId(minOpenWriteId)
```

Comment:
could you please format this for readability

Reply:
Added some formatting. Done.